python/pyarrow/table.pxi

```diff
@@ -5217,6 +5533,13 @@ def record_batch(data, names=None, schema=None, metadata=None):
     if isinstance(data, (list, tuple)):
         return RecordBatch.from_arrays(data, names=names, schema=schema,
                                        metadata=metadata)
+
+    elif isinstance(data, dict):
+        if names is not None:
+            raise ValueError(
+                "The 'names' argument is not valid when passing a dictionary")
+        return RecordBatch.from_pydict(data, schema=schema, metadata=metadata)
```

Comment:
Sorry, while going through your PR and seeing the docstring examples, I was thinking again that we should support a dict in `record_batch`, and quickly did a PR for that before getting to this point of your PR .. (and without reading your description properly). 
I included a bit more docstring changes in https://github.com/apache/arrow/pull/40292, so will keep it or that.

Reply:
I've pulled those changes in given that I'd already had most of it: 

https://github.com/apache/arrow/pull/39506/commits/666b8e0ae7ff2fdbb72860237f14a783d1c06c22